Add maxflow

TonalidadeHidrica opened the pull request.

TonalidadeHidrica commented:
I will add tests and an example code that is submittable to [D - Maxflow](https://atcoder.jp/contests/practice2/tasks/practice2_d), but I will open the PR now.  Waiting everyone for review!

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+            if res == up {
+                break;
+            }
+        }
```
```suggestion
        }
        self.iter[v] = self.graph.g[v].len();
```
This is an important bit of the algorithm that we should not omit. Alternatively we can do
```
self.level[v] = self.graph.g.len();
```

TonalidadeHidrica replied:
You're right.  It is difficult to translate C++ `for`-statements with side effects...

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+    pub fn add_edge(&mut self, from: usize, to: usize, cap: Cap) -> usize {
+        assert!(from < self._n);
+        assert!(to < self._n);
+        assert!(Cap::zero() < cap);
```
```suggestion
        assert!(Cap::zero() <= cap);
```

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+    pub fn add_edge(&mut self, from: usize, to: usize, cap: Cap) -> usize {
+        assert!(from < self._n);
+        assert!(to < self._n);
+        assert!(Cap::zero() < cap);
```
```suggestion
        assert!(Cap::zero() < cap);
        assert_ne!(from, to);
```
or fix `rev` calculation please.

TonalidadeHidrica replied:
Could you teach me why do we need this?  The original documentation says that we can input self-loops to graphs, and I don't think that the code does not work for the situation.  What change do I have to apply to `rev` calculation?

MiSawa replied:
The original implementation has the same bug-ish behavior although both this and the original impl technically "work" as these edges won't be touched. Though it is better to fix, since this bug brakes a fundamental invariant of this graph structure.

The problem is that; when `from == to`, `rev` calculated in L:68 will be the index of the edge added in `L:69`, i.e. itself instead of the reverse edge.
Since `rev` should refer the index of the reversal auxiliary edge, it has to be calculated by
```
 let rev = self.g[to].len() + if from == to { 1 } else { 0 };
```

TonalidadeHidrica replied:
Ah, I've understood.  Fixing.

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+    fn change_edge(&mut self, i: usize, new_cap: Cap, new_flow: Cap) {
+        let m = self.pos.len();
+        assert!(i < m);
+        assert!(Cap::zero() < new_flow && new_flow <= new_cap);
```
```suggestion
        assert!(Cap::zero() <= new_flow && new_flow <= new_cap);
```

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+    pub fn flow(&mut self, s: usize, t: usize) -> Cap {
+        self.flow_with_capacity(s, t, Cap::max_value())
+    }
+    pub fn flow_with_capacity(&mut self, s: usize, t: usize, flow_limit: Cap) -> Cap {
```
What is the reason of this naming? It make sense to name this a `capacity` in some specific context, but here probably doesn't. Is there any reference that this is called capacity?
Maybe `flow_with_limit`, `flow_with_value_limit` or something like that?

TonalidadeHidrica replied:
I have an image of adding new edge from virtual vertex `s'` to vertex `s` with capacity `flow_limit`.  But I agree that `flow_with_limit` may be more intuitive.  I want to ask others' opinion for this point.

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+    pub fn flow_with_capacity(&mut self, s: usize, t: usize, flow_limit: Cap) -> Cap {
+        let n_ = self._n;
+        assert!(s < n_);
+        assert!(t < n_);
```
```suggestion
        assert!(t < n_);
        assert_ne!(s, t);
```

TonalidadeHidrica replied:
I think I should not add this assertion, because the original documentation does not say it disallows the same vertices, and by definition we can return a valid value `flow_limit`.

MiSawa replied:
By definition in the document appendix.html, we should return 0 I believe, which is very confusing as you might have noticed.

> g(v, f) = Sum{...} - Sum{...}

This implies the sum of g(v, f) over all vertices is zero.

> g(v, f) = g(v, f') holds for all vertices v other than s and t.

Since s = t, this also implies g(t, f) = g(t, f').

> g(t,f′)−g(t,f) is maximized under these conditions. It returns this g(t,f′)−g(t,f)

Since g(t, f) = g(t, f'), the returned value should be 0.

TonalidadeHidrica replied:
I finally understood.  It seems to be true that it should return 0.  Doesn't the current code?

TonalidadeHidrica replied:
I tried the original library with a simple example and it returned `MAX_VALUE` instead...

MiSawa replied:
Reported in https://github.com/atcoder/ac-library/issues/5

MiSawa replied:
ACL fixed it :)

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,251 @@
+    fn max_value() -> Self;
+}
+
+impl MfCapacity for i32 {
```
Maybe use macro for these impls?

TonalidadeHidrica replied:
This is a preliminary implementation.  The `internal_type_traits` implementation (#19) provides with more wide variety of features, so I will replace with it once it is merged.

MiSawa commented on src/internal_queue.rs:
```diff
@@ -1 +1,39 @@
+        self.pos = 0;
+    }
+
+    pub(crate) fn pop(&mut self) {
```
Maybe `fn pop(&mut self) -> Option<&T>`?

TonalidadeHidrica replied:
That is an enhancement that is not in the original library, but it seems useful and safe, so I'll add it.  The discussion about `internal_queue` is underway in #13, by the way.

TonalidadeHidrica commented:
@MiSawa Sorry to be late.  I think I added all the assertions you mentioned and fixed all the mistakes.  Now I'm going to add some tests.  Do you have any good idea of test graphs?

MiSawa commented:
> @MiSawa Sorry to be late. I think I added all the assertions you mentioned and fixed all the mistakes. Now I'm going to add some tests. Do you have any good idea of test graphs?

Thank you for being patient about all my nitpicks!

- [This](https://upload.wikimedia.org/wikipedia/commons/9/94/Max_flow.svg) example in Wikipedia has unique optimal solution and very simple. Maybe good as the first test.
- [Here](https://github.com/atcoder/ac-library/blob/master/test/unittest/maxflow_test.cpp)'s unit tests for the original library.
- [Here](https://gist.github.com/MiSawa/47b1d99c372daffb6891662db1a2b686)'s a test case for a common mistake of Dinic algorithm's implementations, although I know the current version is fine.

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,224 @@
 
+impl<Cap> MfGraph<Cap>
+where
+    Cap: Integral,
```
Actually the algorithm doesn't rely on the integrality.... Though it does rely on `Ord`. Not sure what we should do.

TonalidadeHidrica replied:
One possible way is to define a trait dedicated for max-flow capacities, which is my previous implementation.  Refer to [this commit](https://github.com/TonalidadeHidrica/ac-library-rs/commit/c05ac18d2bd8b77f62edba445195d3b66e076690) to see the differences.  However this is a bit like a repetition of `Integral` trait in `internal_type_traits` module, that's why I simply reused it.
At least the original library guarantees only `int` and `ll`, I guess this is already sufficient.

By the way, can this implementation be applied to floating-point capacity?  I mean, don't we need some additional treatment of precision errors?

MiSawa replied:
Good question!
I believe so long as min(x, y) in {x, y} and x-x = 0, the algorithm work fine-ish. e.g. for floating point numbers without NaN / inf, polynomial, etc.
I said fine-ish since although the time complexity doesn't get worse, there are errors that are unavoidable without using multi-precision floating point number. Guaranteeing the accuracy of the returned flow value and flows on edges would be hard. Farthemore since cut is discrete, the min-cut verticis can be different from the correct one although the cut value should be similar.

In competitive programming, I think I have seen some problems that uses polynomial flows (or big-M instead) and floating point numbers (came from binary/ternary search, can be transformed to a search on rational numbers though require a lot more coding).
Though maybe it will be very rare in AtCoder at least, since ACL doesn't support them.

MiSawa reviewed (approved):
LGTM except for missing tests :)

TonalidadeHidrica commented:
About the tests:
* Graphs from Wikipedia is licensed under CC-BY-SA 3.0, so we may use them as long as we mention the reference.
* The license for test cases of original libraries is still yet to be determined, so I'll ask the author if I can use it.
* @MiSawa Under which license can I use your test case?

(I'm not a lawyer, I'm not confident at all..   After all, do I even have to consider the licenses about those small graphs?)

MiSawa commented:
>     * Graphs from Wikipedia is licensed under CC-BY-SA 3.0, so we may use them as long as we mention the reference.

I think CC-BY-SA 3.0 applies to the picture itself, not to the graph instance that is represented by the picture. Though not really sure, and it's safer to mention the origin anyway.

>     * @MiSawa Under which license can I use your test case?

[NYSL](http://www.kmonos.net/nysl/index.en.html)! Added a comment in the gist :)

TonalidadeHidrica commented:
Why does clippy emit warnings in some platforms but not all??

MiSawa commented on src/maxflow.rs:
```diff
@@ -1 +1,325 @@
+
+    #[test]
+    #[allow(clippy::many_single_char_names)]
+    fn test_max_flow_misawa() {
```
You might have done already, but verified locally that it gonna take very long for n=20 and take forever when we deleted L:189 and L:209 as intended :)

TonalidadeHidrica commented:
@qryxip Maybe we can merge this already.
